File: web/src/lib/realtime.ts
import type { ApiClient } from "./api";
import type { ServerEvent } from "./types";

type Listener = (e: ServerEvent) => void;
type Binary = (chunk: ArrayBuffer) => void;

/** The single realtime socket: voice turns, reminder pushes and cross-device sync events. */
export class Realtime {
  private ws: WebSocket | null = null;
  private listeners = new Set<Listener>();
  private binary = new Set<Binary>();
  private stopped = true;
  private attempt = 0;
  private timer: ReturnType<typeof setTimeout> | null = null;
  connected = false;
  onStatus: (connected: boolean) => void = () => {};

  constructor(private client: ApiClient) {}

  start() {
    this.stopped = false;
    this.open();
  }

  stop() {
    this.stopped = true;
    if (this.timer) clearTimeout(this.timer);
    this.ws?.close();
    this.ws = null;
  }

  on(l: Listener) {
    this.listeners.add(l);
    return () => this.listeners.delete(l);
  }

  onBinary(l: Binary) {
    this.binary.add(l);
    return () => this.binary.delete(l);
  }

  send(msg: object) {
    if (this.ws?.readyState === WebSocket.OPEN) this.ws.send(JSON.stringify(msg));
    else throw new Error("not connected");
  }

  sendBinary(data: ArrayBuffer | Blob) {
    if (this.ws?.readyState === WebSocket.OPEN) this.ws.send(data);
  }

  private async open() {
    const url = this.client.wsUrl();
    if (!url || this.stopped) return;
    const ws = new WebSocket(url);
    ws.binaryType = "arraybuffer";
    this.ws = ws;
    ws.onopen = () => {
      this.attempt = 0;
      this.connected = true;
      this.onStatus(true);
    };
    ws.onmessage = (m) => {
      if (typeof m.data === "string") {
        try {
          const evt = JSON.parse(m.data) as ServerEvent;
          this.listeners.forEach((l) => l(evt));
        } catch {
          /* ignore malformed frame */
        }
      } else this.binary.forEach((l) => l(m.data as ArrayBuffer));
    };
    ws.onclose = async (ev) => {
      this.connected = false;
      this.onStatus(false);
      if (this.stopped) return;
      // 4401 = the access token expired: fetch a fresh one through a cheap authenticated call, then reconnect
      if (ev.code === 4401) await this.client.get("/auth/me").catch(() => {});
      const delay = Math.min(15000, 500 * 2 ** this.attempt++) + Math.random() * 300;
      this.timer = setTimeout(() => this.open(), delay);
    };
  }
}

File: mobile/lib/core/realtime.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:web_socket_channel/web_socket_channel.dart';

import 'api_client.dart';

/// The single realtime socket: reminder pushes and cross-device sync events, with reconnect + backoff.
class Realtime {
  Realtime(this._api);
  final ApiClient _api;
  final _events = StreamController<Map<String, dynamic>>.broadcast();
  final _status = StreamController<bool>.broadcast();
  WebSocketChannel? _ch;
  StreamSubscription? _sub;
  Timer? _retry;
  int _attempt = 0;
  bool _stopped = true;

  Stream<Map<String, dynamic>> get events => _events.stream;
  Stream<bool> get status => _status.stream;

  Future<void> start() async {
    _stopped = false;
    await _open();
  }

  Future<void> stop() async {
    _stopped = true;
    _retry?.cancel();
    await _sub?.cancel();
    await _ch?.sink.close();
    _ch = null;
  }

  Future<void> _open() async {
    final uri = await _api.wsUri();
    if (uri == null || _stopped) return;
    try {
      final ch = WebSocketChannel.connect(uri);
      _ch = ch;
      await ch.ready;
      _attempt = 0;
      _status.add(true);
      _sub = ch.stream.listen(
        (data) {
          if (data is String) {
            try {
              _events.add(jsonDecode(data) as Map<String, dynamic>);
            } catch (_) {/* ignore malformed frame */}
          }
        },
        onDone: _lost,
        onError: (_) => _lost(),
      );
    } catch (_) {
      _lost();
    }
  }

  Future<void> _lost() async {
    _status.add(false);
    if (_stopped) return;
    // an expired access token closes the socket with 4401: one authenticated call refreshes it
    try {
      await _api.get('/auth/me');
    } catch (_) {}
    final delay = Duration(milliseconds: min(15000, 500 * pow(2, _attempt++).toInt()) + Random().nextInt(300));
    _retry = Timer(delay, _open);
  }
}
